### DISCUSSION TITLE
[Deployment] Non-inlinable library address should be configured per network

### DISCUSSION THREAD

**Discussion by vmaark** [zkSync-Community-Hub/zksync-developers#763] [Deployments] [answered]
### Team or Project

Treasure

### Environment

Mainnet

### zkSolc Version

zksolc-1.5.7

### zksync-ethers Version

-

### Minimal Reproducible Code or Repo Link

-

### Does this work on other EVMs? (If yes, please list at least 1 of them)

Not applicable

### Description of What Your Contract Does

Non-inlinable libraries have to be deployed separately, and added to the compiler settings in the hardhat config. This is problematic when working with more than 1 deployment targets, let it be a testnet + mainnet, or different chains built with ZK Stack. 

Right now there's no way to set different addresses for different networks. 

Ideally a mapping could be provided with an address for each network, instead of a single address.

**Comment by alhenaai** [comment]
To address the issue of configuring non-inlinable library addresses for different networks in your ZKsync deployment, you can indeed set up a network-specific configuration in your `hardhat.config.ts` file. This allows you to specify different library addresses for each network, facilitating deployments across multiple environments like testnets and mainnet.

Here’s a general approach to achieve this:

1. **Define Library Addresses Per Network**: In your `hardhat.config.ts`, under the `zksolc` settings, you can specify library addresses for each network. This involves setting up a conditional structure where the network name determines the library addresses used.

2. **Example Configuration**:
   ```javascript
   let networkLibraries = {
     mainnet: {
       "contracts/YourLibrary.sol": {
         "YourLibrary": "0xMainnetLibraryAddress"
       }
     },
     testnet: {
       "contracts/YourLibrary.sol": {
         "YourLibrary": "0xTestnetLibraryAddress"
       }
     }
   };

   module.exports = {
     networks: {
       mainnet: {
         url: `https://mainnet.zksync.io`,
         zksolc: {
           version: "zksolc-1.5.7",
           settings: {
             libraries: networkLibraries.mainnet
           }
         }
       },
       testnet: {
         url: `https://testnet.zksync.io`,
         zksolc: {
           version: "zksolc-1.5.7",
           settings: {
             libraries: networkLibraries.testnet
           }
         }
       }
     }
   };
   ```

3. **Deploy Libraries Separately**: Ensure that your libraries are deployed separately on each network and their addresses are updated in the `networkLibraries` object accordingly.

This setup allows you to maintain different library addresses for different networks within the same project configuration, making it easier to manage deployments across various environments.

For more detailed guidance on configuring your Hardhat project for ZKsync, you can refer to the [ZKsync Hardhat integration documentation](https://docs.zksync.io/build/tooling/hardhat/guides/migrating-to-zksync).

If you need further assistance or have more specific requirements, please visit the [ZKsync developers discussions on Github](https://github.com/zkSync-Community-Hub/zksync-developers/discussions) for more in-depth support.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by vmaark** [comment]
I don't think the answer by gleenAI is valid.

**Comment by DusanFedorcak** [comment]
A workaround is to have a separate `hardhat.config.ts` for each deployment target and run all the commands with a selected config. 
In particular, it's useful when you iterate on your own non-linkable libs.

For example, you can have `hardhat.config.local.ts` for local node dev only.
Then, `hardhat --config hardhat.config.local.ts deploy-zksync:libraries` will correctly deploy libraries to the local node and update the local config only.

It effectively removes the need for the `--network` param as you would have one network per config set as the `defaultNetwork`. 

It also deals with local node-related data being committed to a repo, as local config does not need to be committed (and should not be, as each dev would have different library addresses in their config). Unfortunately, a mapping solution proposed in the issue desc. does not account for this.

**Comment by dutterbutter** [comment]
@vmaark updating this to a feature request. This is something we will think about as its becoming quite relevant. Thank you for the post 🙏

**Reply by dutterbutter** [reply, marked as answer]
Opened a feature request here: https://github.com/matter-labs/hardhat-zksync/issues/1531 . Feel free to add any relevant discussion points there. Thanks again for posting 🙏
